### ISSUE TITLE
possible false negative in `unrelated_type_equality_checks`

### ISSUE THREAD

**Issue by pq** [dart-lang/linter#1525]
Follow-up from https://github.com/flutter/flutter/pull/31085#discussion_r275551661.

It's possible we're not properly handling comparisons w/ tear-offs?

https://github.com/dart-lang/linter/blob/046adb919807eedcb749ebbbe9e14860576ed009/lib/src/util/dart_type_utilities.dart#L328-L379

**Comment by pq** [comment]
@srawlins: I see you've been in the `unrelatedTypes` code recently.  I wonder if you have thoughts/opinions?

**Comment by srawlins** [comment]
😶

**Comment by srawlins** [comment]
Yeah it doesn't handle FunctionType. 🙁. I'm happy to land a little fix saying "if one side's type is a FunctionType, and the other side's type is neither a FunctionType nor an InterfaceType, then they're unrelated."

But I can't commit to a wider fix matching parameter types and parameter counts. Does that sound good?

**Comment by srawlins** [comment]
Is there an efficient way to tell if a ClassElement represents a class with a `call()` method? To catch the issue in https://github.com/flutter/flutter/pull/31085#discussion_r275551661, we'd need a test like:

```dart
class C {
  double call() => 7.0;
}

class Matrix4 {
  double determinant() => 3.0;
}

void fn() {
  C c = C();
  Matrix4 m = Matrix4();
  if (m.determinant == c) print('fair'); // OK
  if (m.determinant == 0.0) print('boo'); // LINT
}
```

I doubt that walking through [ClassElement.members](https://pub.dartlang.org/documentation/analyzer/latest/dart_element_element/ClassElement/methods.html) is efficient.

CC @bwilkerson

**Comment by srawlins** [comment]
(Or I could leave callable class instances as a TODO; introduces some false positives but I think fixes way more false negatives.)

**Comment by bwilkerson** [comment]
Iterating through the members of a class element is reasonably performant, but you need to walk the whole hierarchy. There are two methods that will probably do what you want. `ClassElement.lookUpConcreteMethod` will find a concrete implementation of the method, but will return `null` if all of the inherited methods are abstract. If abstract/concrete doesn't matter, use `ClassElement.lookUpMethod`.
